actioncable/test/client_test.rb

```diff
@@ -98,7 +105,26 @@ def with_puma_server(rack_app = ActionCable.server, port = 3099)
       end
     end
   end
+  if defined?(::Iodine)
+    def with_iodine_server(rack_app = ActionCable.server, port = 3099, block)
+      ::Iodine.listen2http(app: rack_app, port: port.to_s, address: "127.0.0.1")
+      ::Iodine.workers = 1 # don't cluster the test
+      ::Iodine.threads = 1 # one for the server another for the task
+      t = Thread.new { ::Iodine.start }
+      block.call(port)
+      ::Iodine.stop
+      t.join
+    end
+  else
+    def with_iodine_server(rack_app = ActionCable.server, port = 3099, block)
+      puts "Iodine testing skipped, unsupported?."
+    end
+  end
 
+  def with_cable_server(rack_app = ActionCable.server, port = 3099, &block)
+    with_puma_server(rack_app, port, block)
+    with_iodine_server(rack_app, port, block)
+  end
```

Comment:
Good question. I really don't know... since they both work, I didn't think about it all that much.

I often debug using `puts` and code comments. For myself, I probably would comment one or the other and re-run the test on my machine... this probably isn't the optimal solution though.

We could add a simple output marker, such as `puts "testing client with Puma/Iodine"` between the tests.

I wonder - it there a way to add / alter the `minitest` output with formatting tags?
 
I'll try and find some way to mark the difference. If you have any ideas, I'd love to read them.

Reply:
I tested a simple solution by adding a simple output line before each server (`"Testing with..."`).

It's easy enough to look back at the logs and see which server/pathway test resulted in a failure.

It's a simple solution and it keeps the client tests a little more DRY.

Reply:
Separate test runs for each server would also address this question.

Reply:
They will, but they will require more code and branches as well as the possibility of repeating some tests that are grouped together but aren't effected by the different servers.

This means either duplicating the blocks to have different line numbers in the error log or factoring the client tests into two separate testing suites in order to isolating the server tests (so we don't repeat the redis tests and the others)...

... but since we need to read the log output anyway (to see the line number whenever a test fails), adding a single output line to the log is, IMHO, a more elegant and versatile solution.

For your consideration.

Reply:
Do you want me to copy the block so the test is unique per server?

Reply:
Logging output does distinguish test failures when you are running the tests on your machine and viewing the logs. It doesn't distinguish test failures well in continuous integration test runs or track whether a test failure is an Iodine issue or a Puma issue or …

A common way to address this is to provide a ClientTests module with the relevant `test_*` methods, then mix that in to two separate test classes, one for each cable server. One of the classes implements `with_cable_server` as `with_puma_server` and the other implements it as `with_iodine_server`. Then other servers may be added by adding new test classes with their own implementations of this one method.

```ruby
module ClientTests
  # … implementation from the current ClientTest class
end

class PumaClientTest < ActionCable::TestCase
  include ClientTests

  private
    def with_cable_server(rack_app = ActionCable.server, port = 3099)
      server = ::Puma::Server.new(rack_app, ::Puma::Events.strings)
      # … etc
    end
end

begin
  require 'iodine'
rescue LoadError
else
  class IodineClientTest < ActionCable::TestCase
    include ClientTests

    private
      def with_cable_server(rack_app = ActionCable.server, port = 3099)
        ::Iodine.listen2http(app: rack_app, port: port.to_s, address: "127.0.0.1")
        # … etc
      end
  end
end
```

Reply:
Thanks, it's a great solution and I didn't think about this option. I'll do that.

Reply:
Done. I hope this matches your expectations.